### PR TITLE
Convert signature verification specs to request specs

### PR DESCRIPTION
**Pull request by ClearlyClaire**
Prerequisite for fixing #18474

I decided to hardcode the signature headers so that:
- the tests are more robust
- other implementers can have a look at examples Mastodon will take as input, together with a keypair to test things with

### CODE REVIEW

**Comment by ClearlyClaire** [comment]
cc @mjankowski since you're working a lot on specs and code style lately, I'd like your feedback on it, especially the long lines, and the routing issues (I used your pattern, but it clears the existing routes)

**Comment by mjankowski** [comment]
> cc @mjankowski since you're working a lot on specs and code style lately, I'd like your feedback on it, especially the long lines, and the routing issues (I used your pattern, but it clears the existing routes)

On the routes/controller - I think what you are doing here is the least bad way to do this. It's just not as smooth in request specs as controller specs where it's more expected to do this. If we find a better way we can update, but at least for now this method matches what's done elsewhere (a CSP spec, maybe?)

On the long lines with the signature values ... the only things I can think of here both from readability and linting perspective would be to put the whole signature value into a heredoc assigned variable, and then do string interpolation when using the value. If the value is purposefully wrong, then just using the raw value makes sense. If the value is calculated from the headers, it might be nice to actually do the calculation here, and/or leave a comment line showing how it was built -- especially if as noted in the PR description, there's some side benefit here to potentially showing others what's being done and what's right/wrong approaches, etc.

**Comment by ClearlyClaire** [comment]
> On the routes/controller - I think what you are doing here is the least bad way to do this. It's just not as smooth in request specs as controller specs where it's more expected to do this. If we find a better way we can update, but at least for now this method matches what's done elsewhere (a CSP spec, maybe?)

It's similar to what you did with the `CSP` spec, yeah. We could also private send to `eval_block` instead of `draw`, but this wouldn't work because of the catch-all `unmatched_route` fallback we have in `config/routes.rb`. Instead, I re-added the one route the code uses.

> On the long lines with the signature values ... the only things I can think of here both from readability and linting perspective would be to put the whole signature value into a heredoc assigned variable, and then do string interpolation when using the value. If the value is purposefully wrong, then just using the raw value makes sense. If the value is calculated from the headers, it might be nice to actually do the calculation here, and/or leave a comment line showing how it was built -- especially if as noted in the PR description, there's some side benefit here to potentially showing others what's being done and what's right/wrong approaches, etc.

Do you have an example on how you'd do it? I'm not sure how a heredoc would help, these values do not have linebreaks.

The idea is that implementers would have values to play with without having to run Mastodon itself.

**Comment by mjankowski** [comment]
> ... but this wouldn't work because of the catch-all `unmatched_route` fallback we have in `config/routes.rb`

That rings a bell from the CSP one -- I recall having to work around that as well.

> Do you have an example on how you'd do it? I'm not sure how a heredoc would help, these values do not have linebreaks.

You'd have to massage it a little bit ... something like this would work:

```
irb(main):001> original = 'Z8ilar3J7bOwqZkMp7sL8sRs4B1FT+UorbmvWoE+A5UeoOJ3KBcUmbsh+k3wQwbP5gMNUrra9rEWabpasZGphLsbDxfbsWL3Cf0PllAc7c1c7AFEwnewtExI83/qqgEkfWc2z7UDutXc2NfgAx89Ox8
DXU/fA2GG0jILjB6UpFyNugkY9rg6oI31UnvfVi3R7sr3/x8Ea3I9thPvqI2byF6cojknSpDAwYzeKdngX3TAQEGzFHz3SDWwyp3jeMWfwvVVbM38FxhvAnSumw7YwWW4L7M7h4M68isLimoT3yfCn2ucBVL5Dz8koBpYf/40w7QidClAw
CafZQFC29yDOg=='
=> "Z8ilar3J7bOwqZkMp7sL8sRs4B1FT+UorbmvWoE+A5UeoOJ3KBcUmbsh+k3wQwbP5gMNUrra9rEWabpasZGphLsbDxfbsWL3Cf0PllAc7c1c7AFEwnewtExI83/qqgEkfWc2z7UDutXc2NfgAx89Ox8DXU/fA2GG0jILjB6Up...
irb(main):002> 
irb(main):003" http_signature_value = <<-SIGNATURE.squish.tr(' ', '')
irb(main):004"   Z8ilar3J7bOwqZkMp7sL8sRs4B1FT+UorbmvWoE+A5UeoOJ3KBcUmbsh+k3wQwbP5gMNUrra9rEWabpa
irb(main):005"   sZGphLsbDxfbsWL3Cf0PllAc7c1c7AFEwnewtExI83/qqgEkfWc2z7UDutXc2NfgAx89Ox8DXU/fA2GG
irb(main):006"   0jILjB6UpFyNugkY9rg6oI31UnvfVi3R7sr3/x8Ea3I9thPvqI2byF6cojknSpDAwYzeKdngX3TAQEGz
irb(main):007"   FHz3SDWwyp3jeMWfwvVVbM38FxhvAnSumw7YwWW4L7M7h4M68isLimoT3yfCn2ucBVL5Dz8koBpYf/40
irb(main):008"   w7QidClAwCafZQFC29yDOg==
irb(main):009> SIGNATURE
=> "Z8ilar3J7bOwqZkMp7sL8sRs4B1FT+UorbmvWoE+A5UeoOJ3KBcUmbsh+k3wQwbP5gMNUrra9rEWabpasZGphLsbDxfbsWL3Cf0PllAc7c1c7AFEwnewtExI83/qqgEkfWc2z7UDutXc2NfgAx89Ox8DXU/fA2GG0jILjB6Up...
irb(main):010> 
irb(main):011> http_signature_value == original
=> true
irb(main):012> 
```

There may be more elegant way there, but that would work.

**Comment by nightpool** [comment]
I wouldn't worry too much about the long lines, but it might be nice to include a simple dev Ruby file or rake task or whatever that shows how you generated the spec values for future maintainability if the Signature headers need to change (e.g. if we change the requests or add a new request type we want to spec)

**Comment by ClearlyClaire** [comment]
I added some methods and assertions on the specs to demonstrate how the signature strings are built. It's not exhaustive because that's not the point of the specs, but this should be enough to illustrate how this works and quickly build specs.

File: spec/requests/signature_verification_spec.rb [hunk 1]
```diff
@@ -0,0 +1,332 @@
...
+        )
+      end
+    end
+  end
+
+  context 'with an HTTP Signature from a known account' do
+    let!(:actor) { Fabricate(:account, domain: 'remote.domain', uri: 'https://remote.domain/users/bob', private_key: nil, public_key: actor_keypair.public_key.to_pem) }
+
+    context 'with a valid signature on a GET request' do
+      let(:signature_header) do
```
Comment (mjankowski) [inline spec/requests/signature_verification_spec.rb:76]: Is it useful to have the long-form `signature` value fully present in the spec? (genuinely asking, not totally clear on the educational intent of these)

I think the added `build_signature_string` is a nice reference improvement, and might be as or more useful than the long calculated strings, which seem to come out of nowhere without seeing how they were built.

For the "wrong value" cases, the specs might better reveal what mistakes are being made by passing in values that are incorrect to the builder method, relative to what is then being used in the request in the spec?

Comment (ClearlyClaire) [reply spec/requests/signature_verification_spec.rb:76]: The idea is that other implementers can look at the code and have reference values to experiment in with their own framework and language without the need to run fully-fledged ruby code.

**Review by renchap — APPROVED** [review]
Looks good to me and I like having more explicit values everywhere.

If you get an approval from @mjankowski (as he is much knowledgable than myself on specs) then go for merge :)

**Review by mjankowski — APPROVED** [review]
Looks solid as move from controller->request spec.

If we get feedback from people on the educational uses, we can always revise style/naming on that front later.

**Pull request merged**